### PR TITLE
[filebeat][azure-blob-storage] - Fixed concurrency & flakey tests issue

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Bug

## What does this PR do?
This PR fixes the concurrency issues present in the azure blob storage input and the flakey tests issue.

## Why is it important?
Concurrent ops were failing at scale and this fix addresses that issue. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Relates #35983

### CODE REVIEW

File: x-pack/filebeat/input/azureblobstorage/job.go [hunk 1]
```diff
@@ -186,16 +189,21 @@ func (j *job) readJsonAndPublish(ctx context.Context, r io.Reader, id string) er
 		// updates the offset after reading the file
 		// this avoids duplicates for the last read when resuming operation
 		offset = dec.InputOffset()
+		var (
+			done func()
+			cp   *Checkpoint
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:194]: completely inconsequential nit that can be ignored; change to reflect assignment and use order.
```suggestion
			cp   *Checkpoint
			done func()
```

File: x-pack/filebeat/input/azureblobstorage/job.go [hunk 2]
```diff
@@ -222,8 +230,11 @@ func (j *job) createEvent(message string, offset int64) beat.Event {
 					},
 				},
 			},
-			"cloud": mapstr.M{
-				"provider": "azure",
+			// Structs are used here in order to save map allocations
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:233]: Why limit this just to this field? The fields above look like they could benefit from the same optimisation. Though I'd suggest that this optimisation be done in a separate, non-bugfix change.

**Review by efd6 — COMMENTED** [review]
Have you confirmed that the tests here fail without the fix?

**Comment by ShourieG** [comment]
@efd6 The concurrency error only happens when using multiple workers at scale with 1000s of records, I'm not sure how we can simulate that without an Integration test.

**Comment by efd6** [comment]
Would it not be possible to make a mock handler that just responds with arbitrary random object data and then run a test with multiple workers pulling from that mock? There does not need to be any comparison of the values, just an absence of a concurrent map-write throw.

**Comment by ShourieG** [comment]
> Would it not be possible to make a mock handler that just responds with arbitrary random object data and then run a test with multiple workers pulling from that mock? There does not need to be any comparison of the values, just an absence of a concurrent map-write throw.

Will try this and update.

**Comment by ShourieG** [comment]
@efd6 So I looked into this and the way the mock will work is call to container -> which gives metadata info wrt all the blobs contained in that container in xml -> call to individual blobs to fetch data. Completely randomising this process means the generation of random metadata in xml and then correlating the blob names generated in this random metadata to the actual blob call path and then generating the random json blob data for that individual blob.

**Comment by ShourieG** [comment]
@efd6 I've update the PR with suitable concurrency tests having random blob generation.

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 3]
```diff
@@ -392,6 +392,132 @@ func Test_StorageClient(t *testing.T) {
...
+			expectedLen: mock.TotalRandomDataSets,
+		},
+		{
+			name: "TestConcurrency_2000_Workers",
+			baseConfig: map[string]interface{}{
+				"account_name":                        "beatsblobnew",
+				"auth.shared_credentials.account_key": "7pfLm1betGiRyyABEM/RFrLYlafLZHbLtGhB52LkWVeBxE7la9mIvk6YYAbQKYE/f0GdhiaOZeV8+AStsAdr/Q==",
+				"max_workers":                         2000,
+				"poll":                                true,
+				"poll_interval":                       "500s",
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:443]: Is this interval different in this case for a reason?

**Review by efd6 — COMMENTED** [review]
Can you confirm that the new tests fail without the fix?

File: x-pack/filebeat/input/azureblobstorage/input_test.go (hunk 3 above)
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:443]: you can ignore this, will revert it to 10s like the others

**Comment by ShourieG** [comment]
> Can you confirm that the new tests fail without the fix?
> 
> I have run the tests with the main branch's code in job.go and state.go, and the tests pass unless `-race` is passed. In the case that `-race` is passed, only the 5000 workers case fails. This failure is consistent, but appears to be purely a timeout effect caused by the added load the race detector imposes; increasing the timeout on the test to 200s allows it to pass.

@efd6 
Nope the new tests seem to be passing even without the fix. The reason is mainly because we are using a stateless input here. But locally I tried even with a stateful input similar to the CEL input tests by passing custom v2 publisher. It seems though the concurrency error is happening deep within the cursor op, specifically the updateCursorOp in the v2 cursor code. I'm not sure how we can mock this behaviour as mocking a complete v2 cursor means mocking the resource and state store packages and the harvester completely.

**Comment by efd6** [comment]
I suggest writing a new beat client implementation (these are small, so it should not be onerous) that mutates the event. This was the cause of the issue, so it should be detectable that way.

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 4]
```diff
@@ -395,24 +386,194 @@ func Test_StorageClient(t *testing.T) {
...
+
+type publisher struct {
+	Channel chan beat.Event
+	mu      sync.Mutex
+	cursors []map[string]interface{}
+}
+
+func (p *publisher) Publish(e beat.Event, cursor interface{}) error {
+	p.mu.Lock()
+	p.Channel <- e
```
Comment (ShourieG) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:550]: @efd6 I've added a custom client to the test but now, even with the concurrency fixes the there seems to be a data race detected here line 550. Any suggestions?

Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:550]: The `Channel` field is being set after the errgroup is started. So you can fix this by moving the assignment up.
```diff
diff --git a/x-pack/filebeat/input/azureblobstorage/input_test.go b/x-pack/filebeat/input/azureblobstorage/input_test.go
index 8d7d49ac9f..f398ab4890 100644
--- a/x-pack/filebeat/input/azureblobstorage/input_test.go
+++ b/x-pack/filebeat/input/azureblobstorage/input_test.go
@@ -507,7 +507,9 @@ func Test_Concurrency(t *testing.T) {
                        v2Ctx, cancel := newV2Context()
                        t.Cleanup(cancel)
                        v2Ctx.ID += tt.name
-                       var client publisher
+                       client := publisher{
+                               Channel: make(chan beat.Event),
+                       }
                        st := newState()

                        var g errgroup.Group
@@ -516,7 +518,6 @@ func Test_Concurrency(t *testing.T) {
                        })
                        timeout := time.NewTimer(100 * time.Second)
                        t.Cleanup(func() { timeout.Stop() })
-                       client.Channel = make(chan beat.Event)
                        var receivedCount int
                wait:
                        for {
```
(note though that `Channel` doesn't need to be exported, I suggest `events`)

Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:550]: I'm not a fan of the channel-based collection pattern that's used here. It's overly complex for what is needed. This is a sketch of what I would do (needs clean-up).
```diff
diff --git a/x-pack/filebeat/input/azureblobstorage/input_test.go b/x-pack/filebeat/input/azureblobstorage/input_test.go
index 8d7d49ac9f..288e17d929 100644
--- a/x-pack/filebeat/input/azureblobstorage/input_test.go
+++ b/x-pack/filebeat/input/azureblobstorage/input_test.go
@@ -507,7 +507,13 @@ func Test_Concurrency(t *testing.T) {
                        v2Ctx, cancel := newV2Context()
                        t.Cleanup(cancel)
                        v2Ctx.ID += tt.name
-                       var client publisher
+                       client := publisher{
+                               stop: func(e []beat.Event) {
+                                       if len(e) >= tt.expectedLen {
+                                               cancel()
+                                       }
+                               },
+                       }
                        st := newState()

                        var g errgroup.Group
@@ -516,38 +522,30 @@ func Test_Concurrency(t *testing.T) {
                        })
                        timeout := time.NewTimer(100 * time.Second)
                        t.Cleanup(func() { timeout.Stop() })
-                       client.Channel = make(chan beat.Event)
-                       var receivedCount int
-               wait:
-                       for {
-                               select {
-                               case <-timeout.C:
-                                       t.Errorf("timed out waiting for %d events", tt.expectedLen)
-                                       cancel()
-                                       return
-                               case got := <-client.Channel:
-                                       _, err = got.Fields.GetValue("message")
-                                       assert.NoError(t, err)
-                                       receivedCount += 1
-                                       if receivedCount >= tt.expectedLen {
-                                               cancel()
-                                               break wait
-                                       }
-                               }
+                       select {
+                       case <-timeout.C:
+                               t.Errorf("timed out waiting for %d events", tt.expectedLen)
+                               cancel()
+                       case <-v2Ctx.Cancelation.Done():
+                       }
+                       g.Wait()
+                       if len(client.events) < tt.expectedLen {
+                               t.Errorf("failed to get all events: got:%d want:%d", len(client.events), tt.expectedLen)
                        }
                })
        }
 }

 type publisher struct {
-       Channel chan beat.Event
+       stop    func([]beat.Event)
+       events  []beat.Event
        mu      sync.Mutex
        cursors []map[string]interface{}
 }

 func (p *publisher) Publish(e beat.Event, cursor interface{}) error {
        p.mu.Lock()
-       p.Channel <- e
+       p.events = append(p.events, e)
        if cursor != nil {
                var c map[string]interface{}
                chkpt, ok := cursor.(*Checkpoint)
@@ -565,6 +563,7 @@ func (p *publisher) Publish(e beat.Event, cursor interface{}) error {

                p.cursors = append(p.cursors, c)
        }
+       p.stop(p.events)
        p.mu.Unlock()
        return nil
 }
```
Note that the test still times out reasonably often when running with `-race` in the 3000 workers case.

**Comment by ShourieG** [comment]
@efd6 I've updated the pr with the suggested changes. For me 3000 workers is not timing out and is passing with the `-race` flag. I've tested this with and without the fix and without the fix the data race occurs and the fix is resolving that issue.

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 5]
```diff
@@ -395,24 +386,193 @@ func Test_StorageClient(t *testing.T) {
...
+				"poll_interval":                       "10s",
+				"containers": []map[string]interface{}{
+					{
+						"name": mock.ConcurrencyContainer,
+					},
+				},
+			},
+			mockHandler: mock.AzureConcurrencyServer,
+			expectedLen: mock.TotalRandomDataSets,
+		},
```
[the remark this answers is not in the document]
Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/input_test.go:424]: It should be possible, but will have to rewrite the test logic a bit. I can take this up as a part of a bigger overhaul I'll be doing in 8.11. Could we approve it for now if it's not a blocker ?

**Comment by ShourieG** [comment]
@efd6 I've updated the test case code.

**Review by efd6 — APPROVED** [review]

**Pull request merged**